File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

map<string, int> m;
int vis[100];

void dfs(string s, int len, string c, char prev) {
    if(s.length() == len) {
        m[s]++;
        return;
    }
    
    for(int i=0; i<c.size(); i++) {
        if(!vis[i] && prev <= c[i]) {
            vis[i] =1;
            dfs(s+c[i], len, c, c[i]);
            vis[i] =0;
        }
    }
}
vector<string> solution(vector<string> orders, vector<int> course) {
    vector<string> ans;
    sort(orders.begin(), orders.end());
    
    for(auto C: course) {
        for(auto O: orders) {
            if(O.size() < C) continue;
            dfs("", C, O, 'A');
        }
        int maxV = -1;
        for(auto M : m) {
            cout << M.first << '\n';
            if(M.second > maxV) maxV = M.second;
        }
        for(auto M : m) {
            if(M.second < 2) continue;
            if(M.second == maxV) ans.push_back(M.first);
        }
        m.clear();
    }
    sort(ans.begin(), ans.end());
    return ans;
}

File: 프로그래머스/2/72411. 메뉴 리뉴얼/메뉴 리뉴얼.java
import java.util.*;
import java.io.*;

class Solution {
    public void dfs(String s, String origin, int idx, Map<String, Integer> map, int goal) {
        if(s.length() == goal) {
            map.put(s, map.getOrDefault(s, 0) + 1);
            return;
        }
        
        for(int i=idx+1; i<origin.length(); i++) {
            dfs(s+origin.charAt(i), origin, i, map, goal);
        }
    }
    public String[] solution(String[] orders, int[] course) {
        String[] answer = {};
        
        for(int i=0; i<orders.length; i++) {
            char[] c = orders[i].toCharArray();
            Arrays.sort(c);
            orders[i] = new String(c);
        }
        
        List<String> ans = new ArrayList<>();
        for(int c : course) {
            
            Map<String, Integer> map = new HashMap<>();
            for(String o : orders) {
                dfs("", o, -1, map, c);
            }
            
            int maxin = -1;
            for(String s : map.keySet()) {
                maxin = Math.max(maxin, map.get(s));
            }
            
            if(maxin >= 2) {
                for(String s : map.keySet()) {
                    if(map.get(s) == maxin) ans.add(s);
                }                    
            }
        }
        
        String[] realans = new String[ans.size()];
        for(int i=0; i<ans.size(); i++) {
            realans[i] = ans.get(i);
        }
    
        
        Arrays.sort(realans);
        return realans;
    }
}